pkg/testing/helm/helm_test.go

```diff
@@ -127,6 +129,104 @@ func (h *HelmTestSuite) TestComponentsFromManifestAfterDryRunHelmInstall(c *C) {
 	c.Assert(len(components), Equals, 7)
 }
 
+// TestSelectedDeploymentAttrFromKanisterHelmDryRunInstall test case does a dry run install of the `kanister` helm chart and validates
+// use cases for `nodeSelector` and `toleration` attributes in the helmValues.yaml. This function is specific to `deployment` resource.
+func (h *HelmTestSuite) TestSelectedDeploymentAttrFromKanisterHelmDryRunInstall(c *C) {
+	nodeSelector := map[string]string{
+		"selector-key": "selector-value",
+	}
+	toleration := []corev1.Toleration{
+		{
+			Key:      "taint-key",
+			Operator: corev1.TolerationOpEqual,
+			Value:    "taint-value",
+			Effect:   corev1.TaintEffectNoSchedule,
+		},
+	}
+
+	var testCases = []struct {
+		testName             string
+		helmValues           map[string]string
+		expectedNodeSelector map[string]string
+		expectedTolerations  []corev1.Toleration
+	}{
+		{
+			testName: "Both nodeSelector and tolerations are present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"nodeSelector.selector-key":   "selector-value",
+				"tolerations[0].key":          "taint-key",
+				"tolerations[0].operator":     "Equal",
+				"tolerations[0].value":        "taint-value",
+				"tolerations[0].effect":       "NoSchedule",
+			},
+			expectedNodeSelector: nodeSelector,
+			expectedTolerations:  toleration,
+		},
+		{
+			testName: "Only nodeSelector is present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"nodeSelector.selector-key":   "selector-value",
+			},
+			expectedNodeSelector: nodeSelector,
+			expectedTolerations:  []corev1.Toleration{},
+		},
+		{
+			testName: "Only tolerations is present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+				"tolerations[0].key":          "taint-key",
+				"tolerations[0].operator":     "Equal",
+				"tolerations[0].value":        "taint-value",
+				"tolerations[0].effect":       "NoSchedule",
+			},
+			expectedNodeSelector: map[string]string{},
+			expectedTolerations:  toleration,
+		},
+		{
+			testName: "Both nodeSelector and tolerations are not present",
+			helmValues: map[string]string{
+				"bpValidatingWebhook.enabled": "false",
+			},
+			expectedNodeSelector: map[string]string{},
+			expectedTolerations:  []corev1.Toleration{},
+		},
+	}
+	for _, tc := range testCases {
+		c.Logf("Test name:%s ", tc.testName)
+		defer func() {
+			h.helmApp.dryRun = false
+		}()
+		c.Log("Installing kanister release from local kanister-operator - Dry run")
+		testApp, err := NewHelmApp(tc.helmValues, kanisterName, "../../../helm/kanister-operator", kanisterName, "", true)
+		c.Assert(err, IsNil)
+
+		out, err := testApp.Install()
+		c.Assert(err, IsNil)
+		components := helm.ComponentsFromManifest(out)
+		c.Assert(len(components), Equals, 7)
+		// Take the deployment component
+		testComponent := components[6]
```

Comment:
this would also break I thin, as soon as we add  a new resource in our helm chart. or if the rendered order is changed somehow.

Reply:
Yes, correct.
So I would do a filter on the component list to pick out the deployment resource, here as that is the resource we are interested in.